Fix check for private key existence before eddsa sign functions

fffonion opened the pull request:
Fix #19524.

Looks like there's no test for cases of RSA and EC, please advice if I need to add test for this patch. Thanks.

A similar fix is also needed for 1.1.1, please let me know if this one looks good so I can backport.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] tests are added or updated

slontis commented:
It should have a test..

slontis commented:
A related question is where EVP_DigestUpdate() should segfault if it is passed NULL as its data..

mattcaswell commented:
> It should have a test..

Yes a test should be there for this. Perhaps in test/evp_extra_test.c

fffonion commented:
I will add test tommorow, thanks for reviewing!

t8m commented on providers/implementations/signature/eddsa_sig.c:
```diff
@@ -161,6 +162,10 @@ int ed25519_digest_sign(void *vpeddsactx, unsigned char *sigret,
         return 0;
     }
+    if (edkey->privkey == NULL) {
+        ERR_raise(ERR_LIB_PROV, EC_R_MISSING_PRIVATE_KEY);
```
This is wrong, you cannot use EC_R with ERR_LIB_PROV. Use `PROV_R_NOT_A_PRIVATE_KEY` instead.

t8m commented on providers/implementations/signature/eddsa_sig.c:
```diff
@@ -198,6 +203,10 @@ int ed448_digest_sign(void *vpeddsactx, unsigned char *sigret,
         return 0;
     }
+    if (edkey->privkey == NULL) {
+        ERR_raise(ERR_LIB_PROV, EC_R_MISSING_PRIVATE_KEY);
```
same here

t8m reviewed (commented):
This is also outside of limits for what we would accept with CLA: trivial. Could you please submit a regular CLA? https://www.openssl.org/policies/cla.html

fffonion commented:
I have addressed the comments and sent CLA to legal. When I was running test locally, the pkey creation always fails with
```
routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:Global default library context, Algorithm (ED25519 : 114), Properties (<null>)
```
It seems the previous test `test_ecx_short_keys` is also failed with same error (although the key passed is not legit as well); I'm wondering if my local setup is wrong or if the test I wrote need to explictly load those algorithms?

t8m commented on test/evp_extra_test.c:
```diff
@@ -4620,14 +4620,24 @@ static int test_ecx_short_keys(int tst)
+    int testresult = 1;
 
 
+    deflprov = OSSL_PROVIDER_load(NULL, "default");
```
This should not be necessary. Unless you have a configuration file that explicitly loads only the base and/or legacy provider which would be fairly broken configuration anyway.

None of the other tests explicitly load the default provider.

fffonion replied:
It seems the there's a test case to use a non-default ctx, where the `null` provider is loaded (https://github.com/openssl/openssl/blob/12b1ae38aa4621a85b00f560a3e828f9a5386d04/test/evp_extra_test.c#L4746), although I'm able to pass all tests after remove that line, but I assume some tests are actually "masked" and only looks to be passed.

t8m replied:
You need to skip the test if testctx is not NULL.

fffonion replied:
That's a good idea, I will update the tests.

t8m commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
+/* Test that trying to sign with a public key errors out gracefully */
+static int test_ecx_not_private_key(int tst)
+{
+#ifdef OPENSSL_NO_EC
+    return 1;
+#else
```
I'd prefer if the whole function was in `#ifndef OPENSSL_NO_EC` and the ADD_ALL_TESTS was also ifndef-ed.

fffonion replied:
fixed

t8m commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
+        case NID_X25519:
+        case NID_X448:
+            /* Test does not apply, skip */
+            return 1;
```
Use `return TEST_skip(.....);`

fffonion replied:
fixed

t8m commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
+    if (keys[tst].pub == NULL)
+        return 1;
+
+    deflprov = OSSL_PROVIDER_load(NULL, "default");
```
This should be again unnecessary.

t8m replied:
Instead skip the test if testcx is not NULL.

t8m commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
+    if (!TEST_ptr(ctx = EVP_MD_CTX_new()))
+        goto err;
+
+    ret = EVP_DigestSignInit(ctx, NULL, NULL, NULL,
+                             pkey);
+
+    if (!TEST_true(ret))
+        goto err;
+
+    if (!TEST_true(EVP_DigestSign(ctx, NULL, &maclen, msg, sizeof(msg))))
+        goto err;
```
In theory, the provider could fail the operation already at the EVP_DigestSignInit() or the EVP_DigestSign() call with NULL signature buffer. I am not sure it is good idea to cement this behavior by the test. Failure of either of these three calls should be accepted as correct.

fffonion replied:
I agree, theoritically those two functions would also fail. But I'm bit concerned if some other failure in those two functions will mask the test, as we don't actually test the error message. How do you think?

t8m replied:
You could call ERR_peek_error() and test that there is the appropriate error reason code on the stack.

t8m commented on test/evp_extra_test.c:
```diff
@@ -4622,11 +4623,15 @@ static int test_ecx_short_keys(int tst)
 
 
+    if (testctx != NULL)
+        return TEST_skip("ecx keys not supported in null provider");
```
The message should be: `test does not work with non-default context`

Hmm, you could actually use EVP_PKEY_new_raw_private_key_ex() function instead and pass the testctx there and the TEST_skip would not be needed then. I am sorry for not realizing this earlier.

fffonion replied:
agreed, that's even better!

t8m commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4652,74 @@ const OPTIONS *test_get_options(void)
+    unsigned char *pubkey;
+    size_t pubkeylen;
+
+    if (testctx != NULL)
+        return TEST_skip("ecx keys not supported in null provider");
```
Same as above.

t8m commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4652,74 @@ const OPTIONS *test_get_options(void)
+
+    /* Check if this algorithm supports public keys */
+    if (keys[tst].pub == NULL)
+        return 1;
```
Again better use `TEST_skip("no public key present");`

t8m commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4652,74 @@ const OPTIONS *test_get_options(void)
+    if (keys[tst].pub == NULL)
+        return 1;
+
+    pubkey = (unsigned char *)keys[tst].pub,
```
Please do not use comma operator unnecessarily.

fffonion replied:
i realized this was a typo after getting confused by this comment, good eyes!

t8m commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4652,74 @@ const OPTIONS *test_get_options(void)
+    if (!TEST_ptr(mac = OPENSSL_malloc(maclen)))
+        goto err;
+
+    if (EVP_DigestSign(ctx, mac, &maclen, msg, sizeof(msg)) != 1)
+        goto check_err;
```
Here I'd use:
```suggestion
    if (!TEST_int_eq(EVP_DigestSign(ctx, mac, &maclen, msg, sizeof(msg)), 0))
        goto err;
```

paulidale commented on test/evp_extra_test.c:
```diff
@@ -4647,6 +4650,73 @@ const OPTIONS *test_get_options(void)
+    size_t pubkeylen;
+
+    switch (keys[tst].type) {
+        case NID_X25519:
```
Style nit, we indent switch statements like this:
```c
    switch (keys[tst].type) {
    case NID_X25519:
    case NID_X448:
        return TEST_skip("signing not supported for X25519/X448");
    }
```

fffonion replied:
fixed!

paulidale reviewed (approved):
Looks good.  Just one indentation style nit to fix and with that, it's approved.

paulidale commented:
@t8m, you probably need to reapprove.

fffonion commented:
Thanks everyone for reviewing!
As this issue also exists in 1.1.1 branch, should I open another PR targeting that branch?

paulidale commented:
1.1.1 is strictly receiving security fixes only at this stage, so probably not.
Unless folks on the @openssl/otc think it should be.

t8m commented:
I agree, this fix is not necessary for 1.1.1 given this is not really security fix.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests.  Once the tests pass it will get moved to 'approval: ready to merge' automatically, alternatively please review and set the label manually.

t8m commented:
Squashed and reworded commit message. Merged to master, 3.1, and 3.0 branches. Thank you for your contribution.
